### PR TITLE
add support for scored named queries

### PR DESCRIPTION
**Pull request by Dharin-shah**
### Description
Opensearch already support labelling the queries, that returns as a list in the returned results, of which query it matched. However one of the use case while doing hybrid search with query text and dense vector is to determine individual scores for each query type. This is very useful in further analysis and building offline model to generate better weights for ranking score. Hence adding this feature that sends the client to add the score for each matched query.

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/pull/6306)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Dharin-shah** [comment]
https://github.com/opensearch-project/OpenSearch/issues/8945
Failing test is flaky

**Comment by Dharin-shah** [comment]
would love a review here @dblock , we at getyourguide would love to have this feature soon via aws

**Comment by Dharin-shah** [comment]
@maintainers if one of you could please review 😬. We are longing for this feature at GetYourGuide

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -632,6 +632,76 @@ public String[] readOptionalStringArray() throws IOException {
...
+     * Map&lt;Integer, String&gt; orderedMap = readOrderedMap(StreamInput::readInteger, StreamInput::readString);
+     * </code></pre>
+     *
+     * @param keyReader The {@link Writeable.Reader} used to read the keys of the map.
+     * @param valueReader The {@link Writeable.Reader} used to read the values of the map.
+     * @return A {@link LinkedHashMap} containing the keys and values read from the data source.
+     *         The map maintains the order in which keys and values were read.
+     * @throws IOException If an I/O error occurs during reading from the data source.
+     */
+    public <K, V> Map<K, V> readOrderedMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader) throws IOException {
```
Comment (reta) [inline libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java:654]: I think we should not have any reliance on order of maps / lists, the caller could use `readMap` and apply whatever ordering (sorting) it wants.

File: server/src/internalClusterTest/java/org/opensearch/search/functionscore/QueryRescorerIT.java [hunk 2]
```diff
@@ -128,7 +128,7 @@ public void testEnforceWindowSize() throws InterruptedException {
                     new QueryRescorerBuilder(
                         functionScoreQuery(matchAllQuery(), ScoreFunctionBuilders.weightFactorFunction(100)).boostMode(
                             CombineFunction.REPLACE
-                        )
+                        ).queryName("hello world")
```
Comment (reta) [inline server/src/internalClusterTest/java/org/opensearch/search/functionscore/QueryRescorerIT.java:131]: We added query name but have not added any assertions?

File: server/src/main/java/org/opensearch/search/SearchHit.java [hunk 3]
```diff
@@ -203,10 +204,14 @@ public SearchHit(StreamInput in) throws IOException {
         sortValues = new SearchSortValues(in);
 
         size = in.readVInt();
-        if (size > 0) {
-            matchedQueries = new String[size];
+        if (in.getVersion().onOrAfter(Version.V_2_12_0)) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/SearchHit.java:207]: We sadly have to go 2 steps, otherwise BWC tests with mixed clusters will fail:

```suggestion
        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
```

File: server/src/main/java/org/opensearch/search/SearchHit.java [hunk 4]
```diff
@@ -559,15 +536,29 @@ public String getClusterAlias() {
         return clusterAlias;
     }
 
-    public void matchedQueries(String[] matchedQueries) {
+    public void matchedQueries(Map<String, Float> matchedQueries) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/SearchHit.java:539]: This is a breaking change of public API (see please this class annotations) for `2.x`, please keep old method as well.

File: server/src/main/java/org/opensearch/search/SearchHit.java [hunk 5]
```diff
@@ -687,12 +678,21 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
             builder.endObject();
         }
         sortValues.toXContent(builder, params);
-        if (matchedQueries.length > 0) {
-            builder.startArray(Fields.MATCHED_QUERIES);
-            for (String matchedFilter : matchedQueries) {
-                builder.value(matchedFilter);
+        if (matchedQueries != null && !matchedQueries.isEmpty()) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/SearchHit.java:681]: The `matchedQueries != null` look redundant, when it could be null?

File: server/src/main/java/org/opensearch/search/SearchHit.java [hunk 6]
```diff
@@ -797,7 +797,25 @@ public static void declareInnerHitsParseFields(ObjectParser<Map<String, Object>,
...
+                        fieldName = p.currentName();
+                    } else if (token.isValue()) {
+                        matchedQueries.put(fieldName, p.floatValue());
+                    }
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                while (p.nextToken() != XContentParser.Token.END_ARRAY) {
+                    matchedQueries.put(p.text(), Float.NaN);
+                }
+            }
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/SearchHit.java:816]: Just to be safe here:

```
         } else {
                    throw new IllegalStateException("expected object or array but got [" + token + "]");
                }
```

**Comment by reta** [comment]
> @maintainers if one of you could please review 😬. We are longing for this feature at GetYourGuide

Apologies for the delay, passed through the change, thank for working on this!

File: server/src/main/java/org/opensearch/search/SearchHit.java (hunk 5 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/SearchHit.java:681]: not really, will remove ambiguity

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java (hunk 1 above)
Comment (Dharin-shah) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java:654]: its more of a convenient method to avoid reordering, and looks a bit cleaner to me, i'd suppose i can remove it but dont really have a super strong opinion. WDYT?

Comment (reta) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java:654]: I would better remove it, the convenience for one-off operation is not sufficient to make it into public API (in my opinion), thank you

File: server/src/main/java/org/opensearch/search/SearchHit.java (hunk 3 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/SearchHit.java]: done

File: server/src/internalClusterTest/java/org/opensearch/search/functionscore/QueryRescorerIT.java (hunk 2 above)
Comment (Dharin-shah) [reply server/src/internalClusterTest/java/org/opensearch/search/functionscore/QueryRescorerIT.java]: added an assertion

File: libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java (hunk 1 above)
Comment (Dharin-shah) [reply libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java]: removed

File: server/src/main/java/org/opensearch/search/SearchHit.java (hunk 4 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/SearchHit.java]: done

File: server/src/main/java/org/opensearch/search/SearchHit.java (hunk 5 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/SearchHit.java]: fixed

File: server/src/main/java/org/opensearch/search/SearchHit.java (hunk 6 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/SearchHit.java:835]: done

File: server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java [hunk 7]
```diff
@@ -101,7 +98,7 @@ public void run() {
             ClusterHealthResponse clusterHealthResponse = client().admin()
                 .cluster()
                 .prepareHealth()
-                .setTimeout(TimeValue.timeValueSeconds(60))
+                .setTimeout(TimeValue.timeValueSeconds(120))
```
Comment (reta) [inline server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java:101]: Why this change?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml [hunk 8]
```diff
@@ -0,0 +1,103 @@
+setup:
+  - skip:
+      version: " - 2.12.0"
```
Comment (reta) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml:3]: I think think `skip` section applies to `matched queries with scores` test case, right?

File: server/src/main/java/org/opensearch/search/SearchHit.java [hunk 9]
```diff
@@ -560,14 +544,38 @@ public String getClusterAlias() {
...
     /**
      * The set of query and filter names the query matched with. Mainly makes sense for compound filters and queries.
      */
     public String[] getMatchedQueries() {
-        return this.matchedQueries;
+        return matchedQueries == null ? new String[0] : matchedQueries.keySet().toArray(new String[0]);
+    }
+
+    /**
+     * @return The score of the provided named query if it matches, {@code null} otherwise.
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/SearchHit.java:568]: Could you please include a comment here (javadocs) that if `include_named_queries_score` is not set, the `Float.NaN` is going to be returned for each named query

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml (hunk 8 above)
Comment (Dharin-shah) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/350_matched_queries.yml:3]: yes

File: server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java (hunk 7 above)
Comment (Dharin-shah) [reply server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java:101]: the test was flaky, got soem timeout errors so increase it to validated and it worked, i can revert if but might be a fix for flakyness

Comment (reta) [reply server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java:101]: It is this one https://github.com/opensearch-project/OpenSearch/issues/9191 ? We could address that separately

Comment (Dharin-shah) [reply server/src/internalClusterTest/java/org/opensearch/indices/recovery/IndexPrimaryRelocationIT.java:101]: yes, okay i can revert it and try to address separately

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 10]
```diff
@@ -71,35 +69,41 @@ public FetchSubPhaseProcessor getProcessor(FetchContext context) throws IOExcept
         for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
             weights.put(
                 entry.getKey(),
-                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE_NO_SCORES, 1)
+                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE, 1)
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:72]: @msfroh would love to hear your opinion: I believe this change could have a visible impact on search (since always calculate scores now). If that is the case, we should be able to pass the state of the `include_named_queries_score` parameter through `FetchContext` and only calculate scores when asked explicitly. What do you think?

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:72]: Thats a good point. I'm also a little concerned about the changes in the processor logic itself. For example, we never check for a two-phase approximation for the scorer (which `Lucene.asSequentialAccessBits` does). Depending on the queries that we're using for matching, this could increase the matching cost. 

(Of course, most scorers that implement `twoPhaseIterator()` use the same implementation for `iterator()`. It looks like `SpanScorer` is an exception, though.)

I'm wondering if it might be worth splitting the `getProcessor` implementation for the case where we want scores?

@Dharin-shah -- thank you very much for implementing this! What approach do you prefer to make sure that the non-scoring behavior doesn't change (functionally and performance-wise) while making sure that scores are available when requested?

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:72]: okay can make some changes, will try and avoid interface breakage, but if it happens i guess its fine

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:72]: so before implementing @msfroh was wondering, can i just use FetchContext to determine if we want to return scores or not?

`  @Override
    public FetchSubPhaseProcessor getProcessor(FetchContext context) throws IOException {
        Map<String, Query> namedQueries = new HashMap<>();
        if (context.parsedQuery() != null) {
            namedQueries.putAll(context.parsedQuery().namedFilters());
        }
        if (context.parsedPostFilter() != null) {
            namedQueries.putAll(context.parsedPostFilter().namedFilters());
        }
        if (namedQueries.isEmpty()) {
            return null;
        }
        Map<String, Weight> weights = new HashMap<>();
        for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
            weights.put(
                entry.getKey(),
                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE, 1)
            );
        }`

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java]: i have added a potential fix, please have a look, however i might suggest that this phase is only executed if there are named queries, so i am not sure how niche the feature usage is but just FYI, might not have that high impact in case if its slow

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java]: @msfroh @reta , if you can please have a look

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 11]
```diff
@@ -71,35 +69,43 @@ public FetchSubPhaseProcessor getProcessor(FetchContext context) throws IOExcept
         for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
             weights.put(
                 entry.getKey(),
-                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE_NO_SCORES, 1)
+                context.includeNamedQueriesScore()
+                    ? context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE, 1)
+                    : context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE_NO_SCORES, 1)
             );
         }
         return new FetchSubPhaseProcessor() {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:77]: I think the @msfroh idea was to have 2 `getProcessor` method implementation, depending on `context.includeNamedQueriesScore() = true | false` , that would be cleaner and easier to maintain vs deciphering conditional logic

File: server/src/main/java/org/opensearch/search/internal/SearchContext.java [hunk 12]
```diff
@@ -305,6 +305,10 @@ public final void assignRescoreDocIds(RescoreDocIds rescoreDocIds) {
 
     public abstract boolean trackScores();
 
+    public abstract SearchContext includeNamedQueriesScore(boolean includeNamedQueriesScore);
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/internal/SearchContext.java:308]: This is a public API, we would need to have a `default` methods here

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java (hunk 11 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:77]: 👍 let me refactor it a bit

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:77]: i do believe we still have to define the score mode based on the includeNamedQueriesScore condition

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:77]: `ScoreMode scoreMode = context.includeNamedQueriesScore() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;`

File: server/src/main/java/org/opensearch/search/internal/SearchContext.java (hunk 12 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/internal/SearchContext.java:308]: sorry wdym by default methods here?

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/internal/SearchContext.java:308]: i will just return "this" i guess to avoid breaking the API

Comment (reta) [reply server/src/main/java/org/opensearch/search/internal/SearchContext.java]: > i will just return "this" i guess to avoid breaking the API

Yeah, empty implementation is fine (meant that by `default` method, sorry for confusion), thank you

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 13]
```diff
@@ -45,63 +15,114 @@
...
         }
+
+        Map<String, Weight> weights = prepareWeights(context, namedQueries);
+
+        return context.includeNamedQueriesScore() ? createScoringProcessor(weights) : createNonScoringProcessor(weights);
+    }
+
+    private Map<String, Query> collectNamedQueries(FetchContext context) {
+        Map<String, Query> namedQueries = new HashMap<>();
+        Optional.ofNullable(context.parsedQuery()).ifPresent(parsedQuery -> namedQueries.putAll(parsedQuery.namedFilters()));
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:39]: Please revert `Optional.ofNullable`, there is no need for it, the old code was just fine

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 14]
```diff
@@ -45,63 +15,114 @@
...
+    }
+
+    private Map<String, Weight> prepareWeights(FetchContext context, Map<String, Query> namedQueries) throws IOException {
         Map<String, Weight> weights = new HashMap<>();
         for (Map.Entry<String, Query> entry : namedQueries.entrySet()) {
-            weights.put(
-                entry.getKey(),
-                context.searcher().createWeight(context.searcher().rewrite(entry.getValue()), ScoreMode.COMPLETE_NO_SCORES, 1)
-            );
+            ScoreMode scoreMode = context.includeNamedQueriesScore() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:47]: 👍

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java (hunk 13 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:39]: pushing

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 15]
```diff
@@ -45,63 +15,114 @@
...
 
             @Override
             public void setNextReader(LeafReaderContext readerContext) throws IOException {
-                matchingIterators.clear();
-                for (Map.Entry<String, Weight> entry : weights.entrySet()) {
-                    ScorerSupplier ss = entry.getValue().scorerSupplier(readerContext);
-                    if (ss != null) {
-                        Bits matchingBits = Lucene.asSequentialAccessBits(readerContext.reader().maxDoc(), ss);
-                        matchingIterators.put(entry.getKey(), matchingBits);
+                setupScorers(readerContext, weights, matchingScorers);
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:59]: Why we have `setupScorers` and `setupMatchingBits`? these methods are used exactly once, we could keep then in the `setNextReader`

File: server/src/main/java/org/opensearch/search/internal/SearchContext.java [hunk 16]
```diff
@@ -305,6 +305,29 @@ public final void assignRescoreDocIds(RescoreDocIds rescoreDocIds) {
 
     public abstract boolean trackScores();
 
+    /**
+     * Determines whether named queries' scores should be included in the search results.
+     * By default, this is set to return false, indicating that scores from named queries are not included.
+     *
+     * @param includeNamedQueriesScore true to include scores from named queries, false otherwise.
+     */
+    public SearchContext includeNamedQueriesScore(boolean includeNamedQueriesScore) {
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/internal/SearchContext.java:314]: 👍

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java (hunk 15 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:59]: yeah went little bit much into the refactoring,  removing those "bit much" methods

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java (hunk 14 above)
Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:47]: good point

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 17]
```diff
@@ -67,41 +33,84 @@ public FetchSubPhaseProcessor getProcessor(FetchContext context) throws IOExcept
...
         }
+
+        Map<String, Weight> weights = prepareWeights(context, namedQueries);
+
+        return context.includeNamedQueriesScore() ? createScoringProcessor(weights) : createNonScoringProcessor(weights);
+    }
+
+    private Map<String, Weight> prepareWeights(FetchContext context, Map<String, Query> namedQueries) throws IOException {
         Map<String, Weight> weights = new HashMap<>();
+        ScoreMode scoreMode = context.includeNamedQueriesScore() ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:44]: 👍

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 18]
```diff
@@ -1,39 +1,9 @@
-/*
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:1]: Sorry but we need to bring header back

File: server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java [hunk 19]
```diff
@@ -45,14 +15,10 @@
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 
-/**
```
Comment (reta) [inline server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:51]: Please bring javadocs back

Comment (Dharin-shah) [reply server/src/main/java/org/opensearch/search/fetch/subphase/MatchedQueriesPhase.java:51]: ofcourse, sorry about that

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
LGMT, thanks @Dharin-shah , @msfroh how does it look to you?

**Comment by Dharin-shah** [comment]
> LGMT, thanks @Dharin-shah , @msfroh how does it look to you?

Thanks, fixing some tests, @msfroh please let me know if this seems okay 🙏

**Review by reta — APPROVED** [review]

**Comment by Dharin-shah** [comment]
https://github.com/opensearch-project/OpenSearch/issues/5329
https://github.com/opensearch-project/OpenSearch/issues/9191
Flaky tests issues reported already

@msfroh possible to take a look and merge this is all good, thanks 🙏

**Review by msfroh — APPROVED** [review]
Sorry for the delay in getting back on this -- it looks good. Thanks a lot @Dharin-shah!

**Comment by reta** [comment]
@Dharin-shah could you please backport to `2.x` manually?

**Comment by andrross** [comment]
@reta @Dharin-shah I've opened the backport in #12427
